### PR TITLE
Reporting Functionality added

### LINKED ISSUE #658: Confirmation of reports generated by volunteers
## Description
As a volunteer, I need to log my shift hours and create a report that needs to be confirmed by the admins. For now, we don't confirm the submitted reports.
So, basically, there is a need to create a **confirm report** functionality for the admins. This report should then be visible to all the admins as an unconfirmed report. Now, as soon as this report is confirmed by at least one of the mentors, we'll move it into the confirmed section of reports.

Currently, the report is not saved. So, after creating a report, we need to save them to database too.

### Update [Required]
- [x] Confirmed and Unconfirmed reports should be displayed separately
- [x] Functionality to confirm reports by admin
- [x] Reports should be saved
- [x] Administrator must be able to view old reports of a volunteer datewise 

## Definition of Done
- [x] All of the required items are completed.
- [x] Approval by 2 mentors.

## Estimation
10 hours

### PR DESCRIPTION
**Pull request by anjali-dhanuka**
# Description
The volunteers can create their reports, administrators can confirm them and also they can view the past reports of the volunteers

Fixes #658 

# Type of Change: 
- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

# How Has This Been Tested?
Volunteer side(volunteer can generate his report):
image

if the report has already been generated:
image

Unconfirmed reports(admin can confirm them):
image

image
Past reports volunteerwise:
image

# Checklist:
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged 

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by anjali-dhanuka** [comment]
Hi @abhishekarya286 @ayushgarg1804 @kamsuri Can you please review the PR.

**Comment by abhishekarya286** [comment]
image

Make the precision of log hours to two decimal places

**Comment by abhishekarya286** [comment]
website is crashing when trying to `log hours` or when I click on `completed shifts tab` whn I have logged in as a volunteer.

image

**Comment by anjali-dhanuka** [comment]
@abhishekarya286  This crash is a separate issue, @Monal5031 has submitted a PR for the same too.

File: vms/administrator/templates/administrator/view_report.html [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+<div class="well well-sm">
+   <a href="{% url 'administrator:approve_report' report.id %}" class="btn btn-success btn-sm">{% trans "Approve Report" %}</a>
+</div>
+{% endif %}
+<div class="well">
+   <legend>{% trans "Report Statistics" %}</legend>
+   <label>
+      <h4><span class="label label-default">{% trans "Total Hours" %}</span></h4>
+   </label>
+   &nbsp&nbsp&nbsp&nbsp&nbsp&nbsp{{total_hours }}
```
Comment (ayushgarg1804) [inline vms/administrator/templates/administrator/view_report.html]: &nbsp";"?
Also can you avoid using nbsp?

File: vms/volunteer/urls.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 # local Django
 from volunteer import views
-from volunteer.views import VolunteerUpdateView, ProfileView, GenerateReportView
+from volunteer.views import VolunteerUpdateView, ProfileView, GenerateReportView, VolunteerHistoryView
```
Comment (ayushgarg1804) [inline vms/volunteer/urls.py]: Long line

File: vms/administrator/templates/administrator/view_report.html (hunk 1 above)
Comment (anjali-dhanuka) [reply vms/administrator/templates/administrator/view_report.html]: @ayushgarg1804 For removing nbsp I have created this issue https://github.com/systers/vms/issues/746

Comment (anjali-dhanuka) [reply vms/administrator/templates/administrator/view_report.html]: Done with inserting ;

File: vms/volunteer/urls.py (hunk 2 above)
Comment (anjali-dhanuka) [reply vms/volunteer/urls.py]: Done

**Comment by anjali-dhanuka** [comment]
@ayushgarg1804 The report gets generated only for logged shifts. Is there still a need to check the future or upcoming ones?
image

**Comment by ayushgarg1804** [comment]
Yes, the report gets generated for logged hours only but currently Volunteer can log hours during or before the shift too, so there must be a check in report generation. Logging might be changed to partial logging after event start, so its advisable if you have a check in report generation.

**Comment by anjali-dhanuka** [comment]
@ayushgarg1804 So should I restrict the report generation to the shifts with logged hrs and end time has passed?

**Comment by ayushgarg1804** [comment]
Yeah

File: vms/shift/services.py [hunk 3]
```diff
@@ -33,7 +34,7 @@ def calculate_duration(start_time, end_time):
     end_delta = datetime.timedelta(
         hours=end_time.hour, minutes=end_time.minute)
     working_hours = (float((end_delta - start_delta).seconds) / 60) / 60
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/shift/services.py]: It was already there so I didn't change it. Do you want me to change it?

File: vms/shift/models.py [hunk 4]
```diff
@@ -75,6 +75,22 @@ class VolunteerShift(models.Model):
...
+class Report(models.Model):
+    total_hrs = models.DecimalField(max_digits=20, decimal_places=2)
+    volunteer_shifts = models.ManyToManyField(VolunteerShift)
+    # confirm_status divides reports into three categories namely
+    # pending:0, approved:1 and rejected:2
+    confirm_status = models.IntegerField(default=0)
+    date_submitted = models.DateField(default=timezone.now)
+    volunteer = models.ForeignKey(Volunteer)
+
+    def get_volunteer_shifts(self):
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/shift/models.py]: Yes, this is the function which gives all the volunteer shifts for which report can be created.

File: vms/shift/services.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (abhishekarya286) [reply vms/shift/services.py]: Yes change that

File: vms/administrator/views.py [hunk 5]
```diff
@@ -18,9 +18,68 @@
...
 from job.services import get_jobs_ordered_by_title
-from shift.services import calculate_total_report_hours, get_administrator_report
+from shift.models import Report
+from shift.services import generate_report, get_report_by_id
 from organization.services import get_organizations_ordered_by_name, get_organization_by_id
 
+class ReportListView(ListView, LoginRequiredMixin):
+   """
+   Returns list of unconfirmed reports
+
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/administrator/views.py]: Yes

File: vms/administrator/templates/administrator/report.html [hunk 6]
```diff
@@ -1,221 +1,40 @@
...
+<h2>{% trans "Manage Reports" %}</h2>
+<hr>
+{% if report_list %}
+<table class="table table-striped table-hover">
+   <thead>
+      <tr>
+         <th>{% trans "Date Submitted" %}</th>
+         <th>{% trans "Volunteer Name" %}</th>
+         <th>{% trans "Organization" %}</th>
+         <th>{% trans "Hours" %}</th>
```
Comment (kamsuri) [inline vms/administrator/templates/administrator/report.html]: I'll suggest to change "Hours" to "Work hours".

File: vms/administrator/templates/administrator/view_report.html [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+<div class="well well-sm">
+   <a href="{% url 'administrator:approve_report' report.id %}" class="btn btn-success btn-sm">{% trans "Approve Report" %}</a>
+</div>
+{% endif %}
+<div class="well">
+   <legend>{% trans "Report Statistics" %}</legend>
+   <label>
+      <h4><span class="label label-default">{% trans "Total Hours" %}</span></h4>
+   </label>
+   &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;{{total_hours }}
```
Comment (kamsuri) [inline vms/administrator/templates/administrator/view_report.html]: Doesn't this needs to be fixed as part of #746 ?

File: vms/volunteer/templates/volunteer/report.html [hunk 8]
```diff
@@ -106,9 +106,7 @@
     {% if report_list %}
         <div class="well">
             <legend>{% trans "Report Statistics" %}</legend>
-            <label><h4><span class="label label-default">{% trans "Total Number of Shifts" %}</span></h4></label>&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp{{ report_list|length }}
```
Comment (kamsuri) [inline vms/volunteer/templates/volunteer/report.html]: Please fix this.

File: vms/volunteer/templates/volunteer/report.html [hunk 9]
```diff
@@ -106,9 +106,7 @@
     {% if report_list %}
         <div class="well">
             <legend>{% trans "Report Statistics" %}</legend>
-            <label><h4><span class="label label-default">{% trans "Total Number of Shifts" %}</span></h4></label>&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp{{ report_list|length }}
-            <br>
-            <label><h4><span class="label label-default">{% trans "Total Hours" %}</span></h4></label>&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp{{ total_hours }}
```
Comment (kamsuri) [inline vms/volunteer/templates/volunteer/report.html]: Same.

File: vms/volunteer/templates/volunteer/report.html [hunk 10]
```diff
@@ -106,9 +106,7 @@
     {% if report_list %}
         <div class="well">
             <legend>{% trans "Report Statistics" %}</legend>
-            <label><h4><span class="label label-default">{% trans "Total Number of Shifts" %}</span></h4></label>&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp{{ report_list|length }}
-            <br>
-            <label><h4><span class="label label-default">{% trans "Total Hours" %}</span></h4></label>&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp{{ total_hours }}
+            <label><h4><span class="label label-default">{% trans "Total Hours" %}</span></h4></label>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;{{ total_hours }}
```
Comment (kamsuri) [inline vms/volunteer/templates/volunteer/report.html]: Same.

File: vms/volunteer/templates/volunteer/view_history.html [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+         </td>
+         <td>
+            <a href="{% url 'administrator:show_report' report.id %}" class="btn btn-info btn-sm">{% trans "View" %}</a>
+         </td>
+      </tr>
+      {% endfor %}
+   </tbody>
+</table>
+{% endblock %}
+
```
Comment (kamsuri) [inline vms/volunteer/templates/volunteer/view_history.html]: Fix #746

File: vms/shift/tests/test_services.py [hunk 12]
```diff
@@ -278,65 +280,22 @@ def test_get_shifts_with_open_slots_for_volunteer(self):
...
-        register(self.v1.id, self.s1.id)
+        # register volunteer for 3 shifts, first: only logged, second: past, third:past and logged shift  
+        register(self.v1.id, self.s.id)
+        register(self.v1.id, self.s1.id)	
         register(self.v1.id, self.s3.id)
 
-        start_time = datetime.time(hour=9, minute=0)
-        end_time = datetime.time(hour=10, minute=0)
-        add_shift_hours(self.v1.id, self.s1.id, start_time, end_time)
+
```
[the remark this answers is not in the document]
Comment (kamsuri) [reply vms/shift/tests/test_services.py]: Fix these extra lines.

File: vms/administrator/templates/administrator/view_report.html (hunk 7 above)
Comment (anjali-dhanuka) [reply vms/administrator/templates/administrator/view_report.html]: Now that PR doesn't exist.

File: vms/volunteer/templates/volunteer/report.html (hunk 10 above)
Comment (anjali-dhanuka) [reply vms/volunteer/templates/volunteer/report.html]: Commented there

File: vms/volunteer/views.py [hunk 13]
```diff
@@ -1,256 +1,316 @@
...
+        organizations_list = get_organizations_ordered_by_name()
+        form = SearchVolunteerForm()
+
+    return render(
+        request, 'volunteer/search.html', {
+            'organizations_list': organizations_list,
+            'form': form,
+            'has_searched': False
+        })
+ations_list': organizations_list,
```
[the remark this answers is not in the document]
Comment (nida) [reply vms/volunteer/views.py]: @anjali-dhanuka address this ?

File: vms/volunteer/tests/test_searchVolunteer.py [hunk 14]
```diff
@@ -242,9 +265,12 @@ def test_volunteer_city_field(self):
         ]
 
         volunteer_2 = create_volunteer_with_details(credentials_2)
+
```
[the remark this answers is not in the document]
Comment (nida) [reply vms/volunteer/tests/test_searchVolunteer.py]: @anjali-dhanuka Formatting ?

File: vms/administrator/templates/administrator/report.html (hunk 6 above)
Comment (anjali-dhanuka) [reply vms/administrator/templates/administrator/report.html]: It is termed as Hours in the original Report already, do you want me to change that also?

**Comment by anjali-dhanuka** [comment]
Hi @abhishekarya286 @nida @kamsuri Done with changes and tests, please review

File: vms/volunteer/templates/volunteer/report.html (hunk 8 above)
Comment (anjali-dhanuka) [reply vms/volunteer/templates/volunteer/report.html]: Fixed

File: vms/volunteer/templates/volunteer/report.html (hunk 10 above)
Comment (anjali-dhanuka) [reply vms/volunteer/templates/volunteer/report.html]: Fixed

File: vms/volunteer/templates/volunteer/report.html (hunk 9 above)
Comment (anjali-dhanuka) [reply vms/volunteer/templates/volunteer/report.html]: Fixed

File: vms/administrator/tests/test_report.py [hunk 15]
```diff
@@ -52,176 +54,101 @@ def login_admin(self):
...
-
-    def create_dataset(self):
-        orgs = Organization.create_multiple_organizations(4)
-
-        test_data = open('test_data.json').read()
-        parameters = json.loads(test_data)
-
-        parameters[0]["org"] = orgs[0]
-        self.register_dataset(parameters[0])
+        self.assertEqual(report_page.element_by_xpath(self.elements.REJECT_REPORT).text, 'Reject')
```
Comment (abhishekarya286) [inline vms/administrator/tests/test_report.py]: Make a function to return this `element_by_xpath(self.elements.REJECT_REPORT).text`.

File: vms/administrator/tests/test_report.py [hunk 16]
```diff
@@ -52,176 +54,101 @@ def login_admin(self):
...
-        parameters[7]["org"] = orgs[3]
-        self.register_dataset(parameters[7])
-
-    """
-    Test giving inconsistent results for log hours and total shifts
-    For log hours: Possibly https://github.com/systers/vms/issues/327 wasn't fixed correctly.
-
-    def test_check_intersection_of_fields(self):
-        self.create_dataset()
+        report_page.element_by_xpath(self.elements.REJECT_REPORT + '//a').click()
```
Comment (abhishekarya286) [inline vms/administrator/tests/test_report.py]: same here

File: vms/administrator/tests/test_report.py [hunk 17]
```diff
@@ -52,176 +54,101 @@ def login_admin(self):
...
+        with self.assertRaises(NoSuchElementException):
+            report_page.element_by_xpath('//table//tbody//tr[1]')
```
Comment (abhishekarya286) [inline vms/administrator/tests/test_report.py]: same

File: vms/administrator/tests/test_report.py [hunk 18]
```diff
@@ -52,176 +54,101 @@ def login_admin(self):
...
+        end=datetime.time(hour=11, minute=0)
+        logged_shift = log_hours_with_details(vol, shift, start, end)
+        create_report_with_details(vol, logged_shift)
+        report_page = self.report_page
         report_page.get_page(self.live_server_url, PageUrls.administrator_report_page)
-        search_parameters_4 = ['', '', 'one', '', '']
-        report_page.fill_report_form(search_parameters_4)
-        self.verify_shift_details('3','2.5')
+        report_page.go_to_view_report_page()
+        self.assertEqual(report_page.element_by_xpath(self.elements.APPROVE_REPORT).text, 'Approve Report')
```
Comment (abhishekarya286) [inline vms/administrator/tests/test_report.py]: Please create appropriate functions

**Comment by abhishekarya286** [comment]
**LGTM**. Thanks @anjali-dhanuka

**Pull request merged**